### ISSUE TITLE
3d scene shows drawing differences compared with osg

### ISSUE THREAD

**Issue by rhabacker** [vsg-dev/VulkanSceneGraph#308]
**Describe the bug**
I get drawing issues with a 3D model compared with osgviewer

**To Reproduce**
1. Build vsg, vsgXchange and vsgExamples
2. start osgviewer with the mentioned 3d model
3. start vsgviewer with the mentioned 3d model
3. Compare the displayed 3d scene of the two applications 

**Expected behavior**
There should be no (major) differences between osg and vsg 

**Screenshots**
See appended screenshots

**Desktop (please complete the following information):**
 - OS: openSUSE Linux
 - Version 15.2

**Additional context**
Converting the osgt file to vsgt with vsgconv and loading that file does not show any differences.
The 3d model has been generated by [FMProfil](https://spie.de/cegit/produkte/freileitungsmanagement/FMProfil)

![Screenshot_20210825_103825](https://user-images.githubusercontent.com/602903/130758067-d8c5f794-5f2e-4304-9fa0-5b27c3caaada.png)
![Screenshot_20210825_104003](https://user-images.githubusercontent.com/602903/130758074-83bc09be-f149-4e2d-ac6f-cb3c5064dc47.png)
[demo.zip](https://github.com/vsg-dev/VulkanSceneGraph/files/7045403/demo.zip)

**Comment by robertosfield** [comment]
I have client work Wednesday through Friday so can't investigate right
away.  Looks like the OSG->VSG conversion is not handling the primitives
types.

Do you have the original file format before it was converted to the OSG?

**Comment by rhabacker** [comment]
The file was generated by dumping the osg 3d scene directly into a osgb file, which I converted to osgt with osgconv and zipped to reduce the file size.

**Comment by robertosfield** [comment]
I have just started looking at the demo.osgt and see the same errors you see.

Looking through the demo.osgt I noticed there are GLSL shaders being used.  As far as I'm aware the present the vsgXchange OSG code doesn't attempt to map shader code across.  I'm not not the author of the OSG->VSG code base so will have to review the code to spot what it does, so it'll be a bit slower to track down things.

Could you provide a bit of background on how the different parts of scene are rendered - what's the balance between fixed function and shader based, and what the intent is behind the shaders usage.

**Comment by rhabacker** [comment]
The shader code embedded in the osgt file
```
19981:   osg::MatrixTransform {
              UniqueID 1455 
              Children 1 {
                osg::Geode {
                  UniqueID 1456 
                  Name "text" 
                  Drawables 1 {
                    osgText::Text {
                      UniqueID 1457 
                      DataVariance STATIC 
                      StateSet TRUE {
                        osg::StateSet {
                          UniqueID 1458 
                          DataVariance STATIC 
                          ModeList 2 {
                            GL_LIGHTING OFF 
                            GL_BLEND ON 
                          }
                          AttributeList 1 {
                            osg::Program {
                              UniqueID 1459 
                              Shaders 2 {
                                osg::Shader {
                                  UniqueID 1460 
                                  Type VERTEX 
                                  ShaderSource 17 {
                                    "$OSG_GLSL_VERSION" 
                                    "$OSG_PRECISION_FLOAT" 
                                    "" 
                                    "$OSG_VARYING_OUT vec2 texCoord;" 
                                    "$OSG_VARYING_OUT vec4 vertexColor;" 
                                    "" 
                                    "void main(void)" 
                                    "{" 
                                    "    gl_Position = gl_ModelViewProjectionMatrix * gl_Vertex;" 
                                    "    texCoord = gl_MultiTexCoord0.xy;" 
                                    "    vertexColor = gl_Color;" 
                                    "" 
                                    "#if !defined(GL_ES) && __VERSION__<140" 
                                    "    gl_ClipVertex = gl_ModelViewMatrix * gl_Vertex;" 
                                    "#endif" 
                                    "}" 
                                    "" 
                                  }
...
```
was created from the [shader code](https://github.com/openscenegraph/OpenSceneGraph/blob/master/src/osgText/shaders/osgText_Text_vert.cpp) of the osgText library for drawing text labels:
```
20341                      Text TRUE "110-/220-kV-Ltg." 
```

**Comment by rhabacker** [comment]
The file demo.osgt file contains a list of osg::Switch nodes each covering a specific set of the scene as shown below:
```
13:         osg::Switch {  "LADMGroup" }
17:         osg::Switch {  "Surface" }
21:         osg::Switch {  "ChangeDetection" }
25:         osg::Switch {  "GELAENDE" }
13193:         osg::Switch {  "GELAENDESOCKEL" }
16312:         osg::Switch {  "daten3d" }
16317:         osg::Switch {  "LAGEPLAN" }
20608:         osg::Switch {  "zusatzgraphik" }
22421:         osg::Switch {  "OBJPG" }
47693:         osg::Switch {  "OBJLI" }
54349:         osg::Switch {  "Objektflaechen" }
59084:         osg::Switch {  "Kreuzung" }
80486:         osg::Switch {  "Masten" }
80491:             osg::Switch {  "Masten_1" }
83845:             osg::Switch {  "Masten_2" }
84570:             osg::Switch {  "Masten_3" }
85295:             osg::Switch {  "Masten_4" }
86001:         osg::Switch {  "Ketten" }
101251:         osg::Switch {  "SEILE" }
109148:         osg::Switch {  "ParMast" }
109153:         osg::Switch {  "ParSeil" }
109158:         osg::Switch {  "PaarKetten" }
109163:         osg::Switch {  "einzellast" }
116537:         osg::Switch {  "abstaende" }
118953:     osg::Switch {  "Selection_Layer" }
```
Most switches contains simple geometry(by using Group, LOD, MatrixTransform, Geode and Geometry osg nodes) as shown below:
```
59084:         osg::Switch {  "Kreuzung" }
59089:             osg::Group {  "FMOSGOBJID:S16_1AFMOSG:KREUZUNGMAST: Station=1052.76 Ordinate=119.68" }
59096:                     osg::Group {  "..\\Symbole3D\\classic\\masten\\mast100.parts[dz3=26.037 dz=24.75]" }
59105:                         osg::Group {  "1" }
59118:                                 osg::Group {  "D:\\2021-08-16-FMPROFIL812 - Daniel-mit-wlog-ohne-rundung-ohne-equal-v4.63-387\\Symbole3D\\classic\\masten\\mastfuss100.obj[ noTriStripPolygons]" }
59109:                             osg::MatrixTransform {  "STREBENIndexedLineSet" }
59122:                                     osg::Geode {  "blassgrÃŒn" }
59185:                                     osg::Geode {  "STREBENIndexedLineSet.001" }
59235:                                     osg::Geode {  "STREBENIndexedLineSet.002" }
59285:                                     osg::Geode {  "STREBEN_1IndexedLineSet" }
59335:                                     osg::Geode {  "STREBEN_1IndexedLineSet.001" }
59385:                                     osg::Geode {  "STREBEN_1IndexedLineSet.002" }
59435:                                     osg::Geode {  "ShapeCylinder" }
60037:                                     osg::Geode {  "ShapeCylinder.001" }
60639:                                     osg::Geode {  "ShapeCylinder.002" }
61241:                                     osg::Geode {  "ShapeCylinder.003" }
61843:                                     osg::Geode {  "ShapeCylinder.004" }
62445:                                     osg::Geode {  "ShapeCylinder.005" }
63047:                                     osg::Geode {  "ShapeCylinder.006" }
63649:                                     osg::Geode {  "ShapeCylinder.007" }
64251:                                     osg::Geode {  "ShapeCylinder.008" }
64853:                                     osg::Geode {  "ShapeCylinder.009" }
65455:                                     osg::Geode {  "ShapeCylinder.010" }
66057:                                     osg::Geode {  "ShapeCylinder.011" }
66659:                                     osg::Geode {  "ShapeCylinder.012" }
67261:                                     osg::Geode {  "ShapeCylinder.013" }
67863:                                     osg::Geode {  "ShapeCylinder.014" }
68465:                                     osg::Geode {  "ShapeCylinder.015" }
69067:                                     osg::Geode {  "ShapeCylinder.016" }
69669:                                     osg::Geode {  "ShapeCylinder.017" }
70271:                                     osg::Geode {  "ShapeCylinder.018" }
70873:                                     osg::Geode {  "ShapeCylinder.019" }
71475:                                     osg::Geode {  "STREBENIndexedLineSet.003" }
71479:                                         osg::Geometry {  "streben" }
```
For showing texts the following structure is used
```
16317:         osg::Switch {  "LAGEPLAN" }
16322:             osg::MatrixTransform {  "LPK" }
19981:             osg::MatrixTransform {  "text" }
19984:                 osg::Geode {  0 { }
16334:                     osg::Geometry {  "glyphTexture" }
19988:                     osgText::Text {  "DepthSortedBin" }
```
As far as I can see, custom shader code only uses the representation of text, all other nodes use the fixed functions

**Comment by robertosfield** [comment]
THanks for the explanation, this helps.

I don't think there is support for Text in the vsgXchange OSG loader, but
now we have vsg::Text this should be possible.  The OSG loader was written
before vsg::Text.

This fixed function subgraphs that should in principle map across just
fine. I looked at the geometry and it looks to be a mixture of quads,
triangles and lines.  I didn't spot anything obvious that might cause
problems,

**Comment by rhabacker** [comment]
Another difference is that the conversation do not convert the name of the osg::Object instance.

The osgt file contains
```
    osg::Group {
      UniqueID 2 
      Name "GroupOfLayers" 
      Children 20 {
        osg::Switch {
          UniqueID 3 
          Name "LADMGroup" 
```
which is converted to
```
  vsg::Node id=2 vsg::Group
  {
    NumUserObjects 0
    children 20
    vsg::Node id=3 vsg::Switch
    {
```
It do not include a `name` attribute nor a key/value pair providing that attribute.

**Comment by robertosfield** [comment]
On Fri, 17 Sept 2021 at 11:39, Ralf Habacker ***@***.***>
wrote:

> Another difference is that the conversation do not convert the name of the
> osg::Object instance.
>
> The osgt file contains
>
>     osg::Group {
>       UniqueID 2
>       Name "GroupOfLayers"
>       Children 20 {
>         osg::Switch {
>           UniqueID 3
>           Name "LADMGroup"
>
> which is converted to
>
>   vsg::Node id=2 vsg::Group
>   {
>     NumUserObjects 0
>     children 20
>     vsg::Node id=3 vsg::Switch
>     {
>
> It do not include a name attribute nor a key/value pair providing that
> attribute.
>
The OSG loader should be considered a first pass, rather than complete. As
the VSG becomes more fully featured the OSG loader will need revisiting to
add features that can now be mapped across.

In the case of the osg::Node name is should be straightforward to utilize
VSG's meta object support i.e. vsg::Object::setValue("name",
osgObj->getName());

Robert.

**Comment by robertosfield** [comment]
I have been investigating the buggy OSG model loading and it looks like the assumption that the osgUtil mesh optimizers would convert quad and quadstrips to triangles isn't valid for you dataset. Vulkan doesn't have any support for quads so will be mishandling these, treating them as triangles.

I'm surprised this issue hadn't been picked up on the test model that we used when the vsgXchange/OSG loader was being written.  This problem points to the need of rewriting the osg::Geometry -> VSG conversion code from scratch.  I can't do this right away, it may be a couple of weeks before I can get to it.

**Comment by robertosfield** [comment]
@rhabacker FYI., I am now following up on some outstanding issues reported. I've addressed the osg::Node::getName() ->vsg::Node::setValue(..) and am now looking a improving the osg::Geometry to VSG equivalents.

**Comment by robertosfield** [comment]
@rhabacker when investigating the problems rendering the model you've provide I came across several VSG serialization bugs that have been introduced with my BufferInfo refactor last month.  I've now fixed these and checked it into VSG master.

These fixes don't fix the problems with the OSG->VSG conversion though.  I've modified the OSG's osgkeyboardmouse example to how me saved selected parts of the scene so I can cherry pick problem bits of geometry and focus bit by bit on my the meshes aren't rendering correctly.  

The first bit of the model I've selected for review is the water.  This osg::Geometry is a bit of an odd one - it's a collection of 6 PrimitiveSet that have a Mode of QUAD_STRP but only 4 indices - so essentially a QUAD.  It's an oddity because rather than six primitives it would be best represented as a single GL_TRIANGLES PrimiritiveSet.  I also see that it has a NormalArray uses BIND_PER_PRIMITIVE_SET but al the values are  0 -0 1, so... a single normal with BIND_OVERALL would have been fine.

I also noticed the grass area was lots of separate geometries where I'd normally expect one.  This is another opportunity to optimize things significantly even within just the OSG realm.  I wouldn't be surprised if could achieve big performance improvements just with better meshing.

However, none of this explains why the conversion of these osg::Geometry is failing so badly.  It may be some of the osgUtil::Optimizer classes that are being invoked to remesh geometries into GL_TRIANGLES.  I haven't yet decided upon a best route through.  My aim for today is to fix this and get the meshes converted across.

How were these meshes created?

**Comment by robertosfield** [comment]
Further investigation that the problem osg::Geometry are not being optimized at all by the osgUtil::IndexMeshVisitor, osgUtil::VertexCacheVisitor, osgUtil::VertexAccessOrderVisitor visitor, instead they are just leaving them intact.  I didn't write these classes so can't say why this off the top of my head.  At a guess I think it might be the per primitive set bindings being flagged as something that can't be handled.

I will take a step back and see if I can come up with a algorithm for mapping these meshes to something that can be done with Vulkan which doesn't have GL_QUADS/GL_QUAD_STRIPS and the awkward bindings.

**Comment by rhabacker** [comment]
Thanks for looking into.

**Comment by robertosfield** [comment]
I have refactored how vsgXchange/osg handles the osg::Geometry primitives to avoid the erroneous quad/triangle tessellation seen in the supplied model. I also added support for meshes with more than 16384 vertices.  

     [https://github.com/vsg-dev/vsgXchange/pull/72](https://github.com/vsg-dev/vsgXchange/pull/72)

During this work I came across the lack of support for points and lines in vsgXchange/osg.  VukanSceneGraph full supports these primitives but the osg->vsg code doesn't yet.  The changes required to fix this are surprisingly widespread as Vulkan requires the GraphicsPipeline to be set up different for points, lines and triangle meshes.  The current vsgXchange/osg implementation is hardwired around one GraphicsPipeline for a range of combination of geometry/state but misses out support for different polygon modes.  Changing this will require extensive changes.

The fixes for the problem tessellation of quad/quad strips that the above fixes provide depend upon a bug fix to the OpenSceneGraph as well as changes to vsgXchange.  I've checked in the required fixes to the 3.6, MeshShaders and master ranches of the OpenSceneGraph.

**Comment by rhabacker** [comment]
> The first bit of the model I've selected for review is the water.
...
> I also noticed the grass area was lots of separate geometries where I'd normally expect one.

Thanks for the pointer - I have made a corresponding request to the FMProfil team for further investigation.

> How were these meshes created?

For the 3D scene, we have written a wrapper around OSG called FMOSG that provides an API for generating geometry from the main application. The cause of the first problem could be therein. For the second case, it is currently not clear where the main application gets the data for the geometries from and if this can be changed.

> I've checked in the required fixes to the 3.6 

Which is https://github.com/openscenegraph/OpenSceneGraph/commit/7505cb46037bf09bd7197cb771e5f6276950c444 for the record.
